lib/posix-process/brk.c

```diff
@@ -0,0 +1,268 @@
+/* SPDX-License-Identifier: BSD-3-Clause */
+/* Copyright (c) 2025, Unikraft GmbH and The Unikraft Authors.
+ * Licensed under the BSD-3-Clause License (the "License").
+ * You may not use this file except in compliance with the License.
+ */
+
+#include <string.h>
+#include <uk/alloc.h>
+#include <uk/event.h>
+#include <uk/print.h>
+#include <uk/prio.h>
+#include <uk/process.h>
+#include <uk/syscall.h>
+#if CONFIG_LIBUKVMEM
+#include <uk/vmem.h>
+#endif /* CONFIG_LIBUKVMEM */
+#include "process.h"
+
+#define HEAP_PAGES			CONFIG_LIBPOSIX_PROCESS_BRK_NPAGES
+#define HEAP_LEN			(HEAP_PAGES << __PAGE_SHIFT)
+
+#if CONFIG_LIBUKVMEM
+static inline int pprocess_brk_init_zone(struct posix_process *pprocess)
+{
+	__vaddr_t va = __VADDR_ANY;
+	int rc;
+
+	rc = uk_vma_map_anon(uk_vas_get_active(), &va,
+			     HEAP_LEN, PAGE_ATTR_PROT_RW, 0, "brk VMA");
+	if (unlikely(rc))
+		return rc;
+
+	pprocess->brk_ctx.base = (void *)va;
+
+	return 0;
+}
+#else /* !CONFIG_LIBUKVMEM */
+static inline int pprocess_brk_init_zone(struct posix_process *pprocess)
+{
+	/* Take a big enough region from the global allocator. */
+	pprocess->brk_ctx.base = uk_palloc(pprocess->_a, HEAP_PAGES);
+	if (unlikely(!pprocess->brk_ctx.base))
+		return -ENOMEM;
+
+	return 0;
+}
+#endif /* !CONFIG_LIBUKVMEM */
+
+int pprocess_brk_init(struct posix_process *pprocess)
+{
+	int rc;
+
+	UK_ASSERT(pprocess);
+	UK_ASSERT(!pprocess->brk_ctx.base);
+
+	rc = pprocess_brk_init_zone(pprocess);
+	if (unlikely(rc)) {
+		uk_pr_err("Failed to create brk space (%lu KiB).\n",
+			  (__u64)HEAP_LEN / 1024);
+		return rc;
+	}
+
+	pprocess->brk_ctx.pos = 0;
+
+	uk_pr_debug("New brk heap region: %p-%p\n",
+		    pprocess->brk_ctx.base,
+		    (__u8 *)pprocess->brk_ctx.base + HEAP_LEN);
+
+	return 0;
+}
+
+#if CONFIG_LIBUKVMEM
+static inline int pprocess_brk_free_zone(struct posix_process *pprocess)
+{
+	return uk_vma_unmap(uk_vas_get_active(),
+			    (__vaddr_t)pprocess->brk_ctx.base, HEAP_LEN, 0);
+}
+#else /* !CONFIG_LIBUKVMEM */
+static inline int pprocess_brk_free_zone(struct posix_process *pprocess)
+{
+	uk_pfree(pprocess->_a, pprocess->brk_ctx.base, HEAP_PAGES);
+	return 0;
+}
+#endif /* !CONFIG_LIBUKVMEM */
+
+int pprocess_brk_free(struct posix_process *pprocess)
+{
+	int rc;
+
+	UK_ASSERT(pprocess->brk_ctx.base);
+
+	rc = pprocess_brk_free_zone(pprocess);
+	if (unlikely(rc))
+		UK_CRASH("Failed to free brk space (%lu KiB).\n",
+			 (__u64)HEAP_LEN / 1024);
+
+	pprocess->brk_ctx.pos = 0;
+	pprocess->brk_ctx.base = __NULL;
+
+	uk_pr_debug("Freed brk heap region: %p-%p\n",
+		    pprocess->brk_ctx.base,
```

Comment:
`pprocess->brk_ctx.base` was set to `__NULL` right above. While we're there, some parts of this file use `NULL`, perhaps the implementation should be consistent and use one or the other.